ibm/service/directlink/resource_ibm_dl_route_report.go

```diff
@@ -0,0 +1,385 @@
+	directLink, err := directlinkClient(meta)
+	if err != nil {
+		return err
+	}
+
+	gatewayId := d.Get(dlGatewayId).(string)
+	createGatewayRouteReportOptionsModel := &directlinkv1.CreateGatewayRouteReportOptions{GatewayID: &gatewayId}
+	routeReport, response, err := directLink.CreateGatewayRouteReport(createGatewayRouteReportOptionsModel)
+	if err != nil {
+		return fmt.Errorf("[DEBUG] Create Direct Link Gateway Route Report (%s) err %s\n%s", gatewayId, err, response)
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s", gatewayId, *routeReport.ID))
+	d.Set(dlRouteReportId, *routeReport.ID)
+
+	isWaitForDirectLinkGatewayRouteReportCompleted(directLink, d.Id(), d.Timeout(schema.TimeoutCreate))
+
+	return resourceIBMDLRouteReportRead(d, meta)
+}
+
+func isWaitForDirectLinkGatewayRouteReportCompleted(client *directlinkv1.DirectLinkV1, ID string, timeout time.Duration) (interface{}, error) {
+	log.Printf("Waiting for direct link route report to be completed for  (%s) ", ID)
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"retry", dlRouteReportPending},
+		Target:     []string{dlRouteReportComplete, ""},
+		Refresh:    isDirectLinkGatewayRouteReportRefreshFunc(client, ID),
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 10 * time.Second,
+	}
+	return stateConf.WaitForState()
+}
+func isDirectLinkGatewayRouteReportRefreshFunc(client *directlinkv1.DirectLinkV1, ID string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+
+		parts, err := flex.IdParts(ID)
+		if err != nil {
+			return nil, "", fmt.Errorf("error getting ID for directlink route report: %s", err)
+		}
+
+		gatewayId := parts[0]
+		routeReportId := parts[1]
+
+		getOptions := &directlinkv1.GetGatewayRouteReportOptions{
+			GatewayID: &gatewayId,
+			ID:        &routeReportId,
+		}
+		routeReport, response, err := client.GetGatewayRouteReport(getOptions)
+		if err != nil {
+			return nil, "", fmt.Errorf("[ERROR] error fetching directlink route report: %s\n%s", err, response)
+		}
+		if *routeReport.Status == "complete" {
+			return routeReport, dlRouteReportComplete, nil
+		}
+		return routeReport, dlRouteReportPending, nil
+	}
+}
+
+func resourceIBMDLRouteReportRead(d *schema.ResourceData, meta interface{}) error {
+	directLink, err := directlinkClient(meta)
+	if err != nil {
+		return err
+	}
+
+	log.Println("[Info] Fetching DL Route Report GW ID:")
+
+	parts, err := flex.IdParts(d.Id())
+	if err != nil {
+		return err
+	}
+
+	gatewayId := parts[0]
+	routeReportId := parts[1]
+
+	log.Println("[Info] Fetching DL Route Report GW ID:", gatewayId, " and Report ID: ", routeReportId)
+
+	getGatewayRouteReportOptionsModel := &directlinkv1.GetGatewayRouteReportOptions{GatewayID: &gatewayId, ID: &routeReportId}
+	report, response, err := directLink.GetGatewayRouteReport(getGatewayRouteReportOptionsModel)
+	if err != nil {
+		log.Println("[WARN] Error fetching DL Route Reports", response, err)
```

Comment:
check for response status == 404.. if 404 setId to "". This helps in lifecycle management.
Refer other resources for reference